omniledger/service/transaction.go

```diff
@@ -3,50 +3,196 @@ package service
+	// a client and still keep the other instructions valid.
+	// Index is relative to the beginning of the clientTransaction.
+	Index int
+	// Length is the total number of instructions in this clientTransaction
+	Length int
+	// Spawn creates a new object
+	Spawn *Spawn
+	// Invoke calls a method of an existing object
+	Invoke *Invoke
+	// Delete removes the given object
+	Delete *Delete
+	// Signatures that can be verified using the darc defined by the objectID.
+	Signatures []darc.Signature
+}
+
+// ObjectID points to an object that holds the state of a contract.
+type ObjectID struct {
+	// DarcID points to the darc controlling access to this object
 	DarcID darc.ID
-	// Nonce: will be concatenated to the darcID to create the key
-	Nonce []byte
-	// Command is object-specific and case-sensitive. The only command common to
-	// all classes is "Create".
+	// InstanceID is taken from the Instruction.Nonce when the Spawn instruction is
+	// sent.
+	InstanceID Nonce
+}
+
+// Slice returns concatenated DarcID and InstanceID.
+func (oid ObjectID) Slice() []byte {
+	return append(oid.DarcID[:], oid.InstanceID[:]...)
+}
+
+// Nonce is used to prevent replay attacks in the instructions and
+type Nonce [32]byte
+
+// Spawn is called upon an existing object that will spawn a new object.
+type Spawn struct {
+	// ContractID represents the kind of contract that needs to be spawn.
+	ContractID string
+	// args holds all data necessary to spawn the new object.
+	Args Arguments
+}
+
+// Invoke calls a method of an existing object which will update its internal
+// state.
+type Invoke struct {
+	// Command is object specific and interpreted by the object.
 	Command string
-	// Kind is only used when the command is "Create"
-	Kind string `proto:"opt"`
-	// Data is a free slice of bytes that can be sent to the object.
-	Data []byte
-	// Signatures that can be verified using the given darcID
-	Signatures []darc.Signature
+	// args holds all data necessary for the successful execution of the command.
+	Args Arguments
+}
+
+// Delete removes the object.
+type Delete struct {
+}
+
+// Argument is a name/value pair that will be passed to the object.
+type Argument struct {
+	// Name can be any name recognized by the object.
+	Name string
+	// Value must be binary marshalled
+	Value []byte
+}
+
+// Arguments is a searchable list of arguments.
+type Arguments []Argument
+
+// Search returns the value of a given argument. If it is not found, nil
+// is returned. So an argument with nil value cannot be distinguished from
+// a missing argument.
+func (args Arguments) Search(name string) []byte {
+	for _, arg := range args {
+		if arg.Name == name {
+			return arg.Value
+		}
+	}
+	return nil
 }
 
 // Hash computes the digest of the hash function
 func (instr Instruction) Hash() []byte {
 	h := sha256.New()
-	h.Write(instr.DarcID)
-	h.Write(instr.Nonce)
-	h.Write([]byte(instr.Command))
-	h.Write([]byte(instr.Kind))
-	h.Write(instr.Data)
+	h.Write(instr.ObjectID.DarcID)
```

Comment:
I've always wondered what to do when doing `h.Write`. Do we check the error or not? `Write` the interface returns an error, but if it's a sha256 then is it still possible to see an error? Right now it's a bit inconsistent, sometimes we check and sometimes we don't.

Reply:
IIRC I got convinced once that `h.Write` will never fail, but I cannot remember the reason anymore.

Reply:
```
type Hash interface {
        // Write (via the embedded io.Writer interface) adds more data to the running hash.
        // It never returns an error.
        io.Writer
```